create STMs based on original topic cfg

bashtanov opened the pull request:
When partition state machines are created topic table may be already way ahead and contain topic config for a newer revision of the topic, or just lack it. This kind of race condition may happen if a topic is deleted shortly after its partition is created. This change is to make sure we use topic config from the same revision as the topic itself.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [x] v24.2.x
- [ ] v24.1.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
cluster/partition: separate STM builder creation from partition start

Refactoring for further changes to STM builder.
Reference type for `xst_state` was changed because defensive moving from
a const subobject in `_raft->start()` call otherwise did not make sense.

bashtanov committed:
cluster/partition_manager: pass whole topic config to create_partition

Instead of indivitual components. Will be needed for correct STM init.

bashtanov committed:
raft/state_machine_manager_builder: pass topic cfg to make_builder_for

For STM factories to use later

bashtanov committed:
cluster/archival_metadata_stm: avoid redundant copies in ctor

bashtanov committed:
STM factories: use topic cfg from params instead of querying topic table

The latter is unreliable, as topic table may be ahead and topic config
may be absent or belong to a future revision of the topic.

bashtanov commented:
/ci-repeat 5
skip-units
dt-repeat=20
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity

bashtanov commented:
/ci-repeat 5
skip-units
dt-repeat=20
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--193bde63-fe38-4dfb-b786-5077ab3a838e-->
<!--TEST RESULTS-->
<!--7ca69f47-3ecc-48e3-830f-50487b3aba16-->
<!--TEST RESULTS-->
<!--a753dbc3-39c0-4900-a7d9-acc95a806c59-->
<!--TEST RESULTS-->
<!--25cb6142-91b7-4293-80d1-6535cb23aaa2-->
<!--TEST RESULTS-->
<!--3465c1de-7e5c-4cd9-8528-2d399f79c0aa-->
<!--TEST RESULTS-->
<!--43868ef3-6293-4c6b-a8f0-5aee9e659d0d-->
<!--TEST RESULTS-->
<!--7ae2babb-8852-4b11-8750-e8947254fdd2-->
<!--TEST RESULTS-->
<!--04657488-2857-449e-bb8f-117b94650655-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64204">build#64204</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.executing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64204#019601f5-ffe7-409f-9351-dbefd0c45bb8 | FLAKY | 20/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64213">build#64213</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64213#01960573-8ea7-4d3b-9642-19cb908af6d5 | FLAKY | 40/41 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64224">build#64224</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/64224#01960f67-e4ee-4027-998a-c83711cf0b52 | FLAKY | 1/2 |
| rptest.tests.tiered_storage_pause_test.TestTieredStoragePause.test_safe_pause_resume.allow_gaps_topic_level=True.allow_gaps_cluster_level=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64224#01960f67-e4ee-4027-998a-c83711cf0b52 | FLAKY | 1/2 |
| rptest.transactions.compaction_e2e_test.CompactionWithRecoveryTest.test_tx_compaction_with_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/64224#01960f67-e4ee-4027-998a-c83711cf0b52 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64375">build#64375</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64375#019620d6-d277-4219-ad86-6bd95535da8f | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64408">build#64408</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64408#019624e2-27d1-4887-908e-9c48de6dbefe | FLAKY | 1/2 |
| rptest.tests.datalake.rest_catalog_connection_test.RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64408#0196253f-3668-457a-9677-b5d125316719 | FLAKY | 1/2 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/64408#0196253f-3668-457a-9677-b5d125316719 | FLAKY | 1/2 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_relaxed_acks.write_caching=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64408#0196253f-366c-414a-836a-bd0c4de44b8d | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64564">build#64564</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/64564#01963a95-8d44-440c-a2f1-9ead9b9269c0 | FLAKY | 1/2 |
| rptest.tests.partition_force_reconfiguration_test.PartitionForceReconfigurationTest.test_basic_reconfiguration.acks=-1.restart=True.controller_snapshots=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64564#01963a90-df98-41fe-be11-d8cd8d8e6c38 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64960">build#64960</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64960#0196639c-df49-42c4-b43b-ac368fd368aa | FLAKY | 1/2 |
| rptest.tests.datalake.disk_budget_test.DatalakeDiskUsageTest.test_idle_finish.num_partitions=10.concurrent_translations=4.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64960#019663e8-1208-4870-b369-c76e52332ec3 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64960#019663e8-120a-4502-8eaf-d97bc1e0710e | FLAKY | 20/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64999">build#64999</a></summary>


| test_id                                                                                                                                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64999#0196674b-1427-4b2d-b903-6a1cf20580ef | FLAKY         | 18/21    |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False                                                                                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64999#0196672f-07ee-4cb7-add0-989da9eab242 | FLAKY         | 13/21    |
</details>
<!--/body-->

bashtanov commented:
/ci-repeat 5
skip-redpanda-build
skip-units
dt-repeat=40
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity

bashtanov commented:
/ci-repeat 5
skip-units
dt-repeat=40
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity

WillemKauf commented on src/v/cluster/archival/archival_metadata_stm.cc:
```diff
@@ -1699,24 +1699,14 @@ bool archival_metadata_stm_factory::is_applicable_for(
-                                 .get_configuration()
-                                 .properties.remote_topic_namespace_override
-                             : std::nullopt;
+    auto cfg = builder.initial_topic_cfg();
```
the `initial_topic_cfg` also comes from `_topics.local().get_topic_metadata_ref()`, but in `controller_backend`. Can you explain why we haven't just moved the window for the race condition slightly with this change?

(I assume it's because we haven't opened the topic up for `alter-config` exposure yet, but if you could spell it out explicitly that would be nice)

bashtanov replied:
We are not overly worried about using a stale or a too fresh config for the topic, as it is going to be updated later on anyway. What we are more worried about is using a config of a past or future revision of the topic, i.e. a recreated topic with the same name.

Controller backend is meant to be free from race conditions like this, but it looks like it has a similar bug. I'll address it, thanks for finding it.

ztlpn commented on src/v/raft/state_machine_manager.h:
```diff
@@ -238,7 +248,12 @@ class state_machine_manager_builder {
+    }
+
 private:
+    const cluster::topic_configuration* _initial_topic_cfg;
```
default initializer needed

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64375-->
<!--8eed922b-0fbf-4df1-a485-6bdf8ac5798d-->
<!--title-->
## Retry command for [Build#64375](https://buildkite.com/redpanda/redpanda/builds/64375)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":true}
```
<!--/body-->

WillemKauf commented on src/v/cluster/controller_backend.cc:
```diff
@@ -1295,21 +1297,13 @@ ss::future<std::error_code> controller_backend::create_partition(
-            // topic was already removed, do nothing
-            co_return errc::success;
-        }
+        vassert(topic_md, "topic metadata disappeared for {}", ntp);
```
Changing this from an expected and handled case to a `vassert()` is scary on first glance. How confident are we in this?

bashtanov replied:
[We are fairly confident](https://github.com/redpanda-data/redpanda/pull/25781#discussion_r2045623527l) with things as they are, but [there was a request from Nicolae to make it more robust on subsequent code changes](https://github.com/redpanda-data/redpanda/pull/25781#discussion_r2045123300)

mmaslankaprv commented on src/v/raft/state_machine_manager.h:
```diff
@@ -238,7 +248,12 @@ class state_machine_manager_builder {
         return {raft, std::move(_stms), _sg};
     }
 
+    const cluster::topic_configuration* initial_topic_cfg() const {
```
Raft should never depend on the types from `cluster` submodule. Why generic `raft::state_machine_manager` has to be aware of the topic ? I think it should not be

bashtanov replied:
It does not depend on the type as such, as the type is incomplete.

The assumption of a state machine factory being capable to create state machines from consensus instance only is not correct and leads to race conditions we are fixing here. We need a way to pass topic config to the state machine constructor. 

The only alternative I can see would be to add a parameter to `state_machine_factory::create` method and pass it to all STMs whether they need it or not. Would you prefer it?

Also, whether we do the above change or not, I can wrap topic config pointer into a less specifically named class, e.g. `cluster::stm_init_params` to avoid adding more logic into this area should we need to pass something else in future. Do you think it's a good idea?

bashtanov committed:
cluster/state_machine_registry: supply instance context to STM ctors

STMs were created using only raft, group id and STM factory until this
change. As STM factories are shared between STM instances of the same
type used for different partitions, no partition-specific info could be
passed.

This change introduces cluster::stm_instance_config struct which holds
specific info needed for STM constructors. Currently it's only a pointer
to topic config.

We do not use it in individual STM factories yet.

bashtanov committed:
cluster/controller_backend: fix race when marking properties reconciled

Make sure that topic table access in `reconcile_ntp_step` happens in the
same scheduling segment and is thus consistent.

bashtanov committed:
cluster/controller_backend/create_partiton: query topic tbl consistently

Make this coroutine
1) never query topic table itself
2) only use references to data queried from topic table by caller before
the first scheduling point

Calling `create_vnode_set` from replica revision map (which comes from
topic_table::partition_replicas_view) may be some redundant work in case
partition creation is aborted. But it is rare, typically the partition
still needs to be created. And making a copy of replica revision map
does not look cheaper, as absl::flat_hash_map copy constructor is not a
simple memcpy.

bashtanov committed:
STM factories: do not depend on topic table

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64960-->
<!--4426981e-3636-4fe9-b941-65b1151fc079-->
<!--09b17cd9-cb9d-4c86-8875-9569191cab57-->
<!--title-->
## Retry command for [Build#64960](https://buildkite.com/redpanda/redpanda/builds/64960)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/datalake/disk_budget_test.py::DatalakeDiskUsageTest.test_idle_finish@{"cloud_storage_type":1,"concurrent_translations":4,"num_partitions":10}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":false,"mixed_versions":false,"with_chunked_compaction":false,"with_iceberg":true,"with_tiered_storage":true}
```
<!--/body-->

ztlpn commented on src/v/cluster/state_machine_registry.h:
```diff
@@ -16,6 +16,15 @@
+    /* Any context for STM instance creation */
+
+    // can be null for controller
+    const cluster::topic_configuration* initial_topic_cfg;
```
default initializer? just in case

bashtanov replied:
I created an explicit constructor instead so that this pointer value is always specified.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64999-->
<!--1d7a30c3-c745-4697-bb47-872329c7d730-->
<!--fe0e41cb-e4ce-4db5-985c-e331b9f3694d-->
<!--title-->
## Retry command for [Build#64999](https://buildkite.com/redpanda/redpanda/builds/64999)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":true,"mixed_versions":false,"with_chunked_compaction":true,"with_iceberg":true,"with_tiered_storage":false}
```
<!--/body-->

bashtanov commented:
bazel fixture test failure [is unrelated](https://github.com/redpanda-data/redpanda/pull/25916)

bashtanov commented:
ducktape failure is https://redpandadata.atlassian.net/browse/CORE-9747

bashtanov commented:
@mmaslankaprv could you merge if you don't have any concerns?

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25728-v24.2.x-633 remotes/upstream/v24.2.x
git cherry-pick -x 2e6178b102 2b304b36f4 088295b6d6 09c2a4a8d5 b25fa8ed62 5b49a6ce00 b349eb96a4 3964846e3f a0d13bdfc9
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14659543951).

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25728-v24.3.x-818 remotes/upstream/v24.3.x
git cherry-pick -x 2e6178b102 2b304b36f4 088295b6d6 09c2a4a8d5 b25fa8ed62 5b49a6ce00 b349eb96a4 3964846e3f a0d13bdfc9
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14659543753).

dotnwat reviewed (commented):
nice catch
